api/v1alpha4/openstackmachine_types.go

```diff
@@ -61,6 +61,9 @@ type OpenStackMachineSpec struct {
 	// UUID, IP address of a port from this subnet will be marked as AccessIPv4 on the created compute instance
 	Subnet string `json:"subnet,omitempty"`
 
+	// create and assign additional ports to instances
+	Ports []PortOpts `json:"ports,omitempty"`
```

Comment:
We only need VNICType actually. I try to find a light solution.
In my understanding, SR-IOV setting is done per compute node, but real use case is that create network for SR-IOV and create port for it. How about adding VNICType to NetworkParam? At least we can use existing network with SR-IOV.



we can add only VNICType. Because only VNICType is needed for SR-IOV.

In my understanding, OpenStack can change SR-IOV setting per compute node, but Cluster API can change the setting per MachineDeployment in availability zone.




Since we need only VNICType for SR-IOV, it seems

Reply:
+1 Since Network param and Subnet Param are no longer nested in v1aplha4 api, it would also make sense to be able to set the vnicType in the subnet, in case that is a user's preferred way to set an interface.

Reply:
I'm basically fine with whatever option you get consensus on :)
But just asking. We didn't change NetworkParam and SubnetParam from v1alpha4 and as far as I can see they are still nested. Did I miss something that they are not nested anymore ? :)

Reply:
I could be misunderstanding, but: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/e1a13286736fd151106ddadd730a250713efde71/api/v1alpha4/openstackmachine_types.go#L57-L62

It looks like they are their own isolated api entry. Its very possible that I am incorrect though :)

Reply:
Ah now I get what lead to this assumption. This Subnet is only for the AccessIPv4 "selection". (it was introduced in https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/756)

SubnetParam is still nested.

https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/e1a13286736fd151106ddadd730a250713efde71/api/v1alpha4/types.go#L61-L70

As mentioned elsewhere we have to rethink our design / CRD sooner or later as it's not that thought out / consistent.

Reply:
ah ok, that makes sense now. thanks for explaining!